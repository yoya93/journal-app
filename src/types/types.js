export const types = {
  login: "[Auth] Login",
  logout: "[Auth] Logout",

  uiStartLoading: "[UI] Start loading",
  uiFinishLoading: "[UI] Finish loading",

  uiSetError: "[UI] Set Error",
  uiRemoveError: "[UI] Remove Error",
};
