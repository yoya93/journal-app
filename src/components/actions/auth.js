import { types } from "../../types/types";
import { firebase, googleAuthProvider } from "../../firebase/firebase-config";

export const startLoginEmailPassword = (email, passsword) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(1234, "Yoya"));
    }, 3500);
  };
};

export const startGoogleLogin = (email, passsword) => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});
