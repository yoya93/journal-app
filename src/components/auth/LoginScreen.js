import React from "react";
import { Link } from "react-router-dom";

import { useForm } from "../../hooks/useForm";
import { startLoginEmailPassword, startGoogleLogin } from "../actions/auth";
import { useDispatch } from "react-redux";

export const LoginScreen = () => {
  const [formValues, handleInputChange] = useForm({
    email: "@espainosa",
    password: "espinosa",
  });

  const dispatch = useDispatch();

  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin(email, password));
  };

  return (
    <>
      <h1 className="auth__title">Login</h1>

      <form onSubmit={handleLogin}>
        <input
          type="text"
          autoComplete="off"
          placeholder="Email"
          name="email"
          value={email}
          className="auth__input"
          onChange={handleInputChange}
        />
        <input
          type="password"
          autoComplete="off"
          placeholder="password"
          name="password"
          value={password}
          className="auth__input"
          onChange={handleInputChange}
        />
        <button className="btn btn-primary btn-block" type="submit">
          login
        </button>
        <div className="auth__social-networks">
          <p>Login with social networks</p>
          <div className="google-btn">
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p onClick={handleGoogleLogin} className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link className="link" to="/auth/register">
          Create new account
        </Link>
      </form>
    </>
  );
};
