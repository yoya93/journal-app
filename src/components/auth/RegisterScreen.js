import React from "react";
import { Link } from "react-router-dom";
import validator from "validator";
import { useDispatch, useSelector } from "react-redux";

import { useForm } from "../../hooks/useForm";
import { setError, removeError } from "../actions/ui";
import { startRegisterWithEmailPasswordName } from "../actions/auth";

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { msgError } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    name: "Yoya",
    email: "@espainosa",
    password: "1234",
    password2: "1234",
  });

  const { name, email, password, password2 } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();
    if (isFormValid()) {
      dispatch(removeError());
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError("name is required"));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError("Email is not valid"));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          "Password should be at least 6 characters and match each other"
        )
      );

      return false;
    }

    return true;
  };

  return (
    <div className="animate__animated animate__fadeIn animate__faster">
      <h1 className="auth__title">Register</h1>

      <form onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <input
          type="text"
          autoComplete="off"
          placeholder="Name"
          name="name"
          value={name}
          onChange={handleInputChange}
          className="auth__input"
        />
        <input
          type="text"
          autoComplete="off"
          placeholder="Email"
          name="email"
          value={email}
          onChange={handleInputChange}
          className="auth__input"
        />
        <input
          type="password"
          autoComplete="off"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
          className="auth__input"
        />
        <input
          type="password"
          autoComplete="off"
          placeholder="Confirm password"
          name="password2"
          value={password2}
          onChange={handleInputChange}
          className="auth__input"
        />
        <button type="submit" className="btn btn-primary btn-block mb-5">
          login
        </button>

        <Link className="link " to="/auth/login">
          Already registered?
        </Link>
      </form>
    </div>
  );
};
