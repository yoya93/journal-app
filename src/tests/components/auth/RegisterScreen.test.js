import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store"; //ES6 modules
import thunk from "redux-thunk";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom"; // simulo que estoy dentro de una router para ke se pueda usar el link

import {
  startGoogleLogin,
  startLoginEmailPassword,
} from "../../../components/actions/auth";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { types } from "../../../types/types";

const middlewares = [thunk];

// jest.mock("../../../components/actions/auth", () => ({
//   // se le hace un tranking a la func para saber si es disparada en el component
//   // pero como se llamam dentro de un dispatcher es necesario seguir tambien al store.dispatcher
//   startGoogleLogin: jest.fn(),
//   startLoginEmailPassword: jest.fn(),
// }));

const mockStore = configureStore(middlewares);
const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null,
  },
};

let store = mockStore(initState);
//store.dispatch = jest.fn();// en este test tenemos ke kitar el trakimg al dispatch del store porke todas las acciones son asincronas
//por lo tanto no necesitamos simular el dispatch
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
);

describe("Pruebas en <RegisterScreen/>", () => {
  test("debe de mostrarse correctamente el componente ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de disparar la action  de email vacio ", () => {
    const emailField = wrapper.find('input[name="email"]');

    emailField.simulate("change", {
      target: {
        value: "",
        name: "email",
      },
    });

    const e = {
      preventDefault: () => {},
    };
    wrapper.find("form").prop("onSubmit")(e);
    const action = store.getActions();

    expect(action[0]).toEqual({
      type: types.uiSetError,
      payload: "Email is not valid",
    });
  });

  test("debe de mostrar el msg de email vacio ", () => {
    const initState = {
      auth: {},
      ui: {
        loading: false,
        msgError: "email error",
      },
    };

    const store = mockStore(initState);

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".auth__alert-error").exists()).toBe(true);
    expect(wrapper.find(".auth__alert-error").text().trim()).toBe(
      initState.ui.msgError
    );
  });
});
