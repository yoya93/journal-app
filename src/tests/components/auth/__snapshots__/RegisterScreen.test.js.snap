// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen/> debe de mostrarse correctamente el componente  1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h1
    className="auth__title"
  >
    Register
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Yoya"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="@espainosa"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="1234"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="1234"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      login
    </button>
    <a
      className="link "
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</div>
`;
